uwblueprint/planet-read
# Remove user from story translation mutation

## Notion ticket link



[Remove user from story translation mutation](https://www.notion.so/uwblueprintexecs/Remove-user-from-story-translation-mutation-eb2cf484e02b49aeb9545f52f4c6ba0c)



## Implementation description

- add mutation to remove user from story translation
- if reviewer, set `reviewer_id`=None
- if translator, soft delete translation
- added basic unit tests for `story_service`



## Steps to test

1. Verify that database is up to date and seeded
2. Verify that you cannot make [this request](http://localhost:5000/graphql?variables=null&query=mutation%20%7B%0A%20%20removeUserFromStoryTranslation(storyTranslationId%3A%2013%2C%20userId%3A%201)%20%7B%0A%20%20%20%20ok%0A%20%20%7D%0A%7D) if unauthorized
3. Login as admin; get accessToken and set as header in Altair
```
mutation {
  login(email:"planetread+angelamerkel@uwblueprint.org",password:"changeme"){
    accessToken
  }
}
```
4. Remove reviewer from story translation:
```
mutation {
  removeUserFromStoryTranslation(storyTranslationId: 13, userId: 4) {
    ok
  }
}
```
Check that in the database, the `reviewer_id` is now `NULL`: `			docker exec -it planet-read_db_1 mysql -u root -proot -e "USE planet-read; SELECT * FROM story_translations WHERE id=13;"`
5. Try to remove reviewer again; verify that exception is raised: `Error. User is not a translator or reviewer of this story translation.`
6. Remove translator from story_translation:
```
mutation {
  removeUserFromStoryTranslation(storyTranslationId: 13, userId: 1) {
    ok
  }
}
```
Check that in the database, the story translation is soft deleted (`is_deleted`=1): `docker exec -it planet-read_db_1 mysql -u root -proot -e "USE planet-read; SELECT * FROM story_translations_all WHERE id=13;"`
7. Try to remove translator from story_translation again; verify that exception is raised: `Error. Story translation does not exist`
8. Verify that the tests pass



## What should reviewers focus on?
- `backend/python/app/services/implementations/story_service.py` / `backend/python/app/tests/services/test_story_service.py`
- do the tests make sense?
- story_service is getting quite long--is there a logical way to break some of the functions out (e.g. story vs. story_translation)

## Checklist
